src/variational/VariationalInference.jl

```diff
@@ -0,0 +1,171 @@
+module Variational
+
+using ..Core, ..Core.RandomVariables, ..Utilities
+using Distributions, Bijectors
+using ProgressMeter, LinearAlgebra
+using ..Turing: PROGRESS
+using ..Turing: Model, SampleFromPrior, SampleFromUniform
+using ..Turing: Turing
+using Random: AbstractRNG
+
+using ForwardDiff
+using Flux.Tracker, Flux.Optimise
+
+import ..Core: getchunksize, getADtype
+
+export
+    vi,
+    ADVI,
+    ELBO
+
+
+abstract type VariationalInference{AD} end
+
+getchunksize(::T) where {T <: VariationalInference} = getchunksize(T)
+getchunksize(::Type{<:VariationalInference{AD}}) where AD = getchunksize(AD)
+getADtype(alg::VariationalInference) = getADtype(typeof(alg))
+getADtype(::Type{<: VariationalInference{AD}}) where {AD} = AD
+
+abstract type VariationalObjective end
+
+const VariationalPosterior = Distribution{Multivariate, Continuous}
+
+"""
+    rand(vi::VariationalInference, num_samples)
+
+Produces `num_samples` samples for the given VI method using number of samples equal to `num_samples`.
+"""
+function rand(vi::VariationalPosterior, num_samples) end
+
+"""
+    objective(vi::VariationalInference, q::VariationalPosterior, model::Model, args...)
+
+Computes the variational objective to be optimized for a given VI method.
+"""
+function objective(vi::VariationalInference, q::VariationalPosterior, model::Model, num_samples) end
```

Comment:
Similar here.

Reply:
Same here

Reply:
Similar to `rand`. I'm now of the opinion that we don't need to force the user to make an `objective` function, but instead allow any arbitrary _callable_ work as an objective.

Remove it for now?